airgun/views/ansible_variable.py

```diff
@@ -0,0 +1,91 @@
+from widgetastic.widget import Checkbox
+from widgetastic.widget import Text
+from widgetastic.widget import TextInput
+from widgetastic_patternfly import BreadCrumb
+
+from airgun.views.common import BaseLoggedInView
+from airgun.views.common import SatTable
+from airgun.views.common import SearchableViewMixin
+from airgun.widgets import FilteredDropdown
+from airgun.widgets import Pagination
+from airgun.widgets import SatSelect
+
+
+class AnsibleVariablesView(BaseLoggedInView, SearchableViewMixin):
+    """Main Ansible Variables view."""
+
+    title = Text("//h1[contains(., text()='Ansible Variables')")
+    new_variable = Text("//a[contains(@href, '/ansible/ansible_variables/new')]")
+    total_variables = Text("//span[@class='pagination-pf-items-total']")
+    table = SatTable(
+        './/table',
+        column_widgets={
+            'Name': Text("./a"),
+            'Role': Text("./a"),
+            'Type': Text("./a"),
+            'Imported?': Text("./a"),
+            'Actions': Text("./a[contains(., text()='Delete')]"),
+        },
+    )
+    pagination = Pagination()
+
+    @property
+    def is_displayed(self):
+        return self.browser.wait_for_element(
+            self.title, exception=False
+        ) is not None and self.browser.url.endswith('ansible_roles')
+
+
+class NewAnsibleVariableView(BaseLoggedInView):
+    """View while creating a new Ansible Variable"""
+
+    breadcrumb = BreadCrumb()
+    key = TextInput(id='ansible_variable_key')
+    description = TextInput(id='ansible_variable_description')
+    ansible_role = FilteredDropdown(id='ansible_variable_ansible_role_id')
+    override = Checkbox(id='ansible_variable_override')
+
+    # Accessing all widgets except the ones above requires that the `override` checkbox is filled
+    parameter_type = SatSelect(id='ansible_variable_parameter_type')
+    default_value = TextInput(id='ansible_variable_default_value')
+    hidden_value = Checkbox(id='ansible_variable_hidden_value')
+    expand_input_validator = Text("//h2[@class='expander collapsed']")
+    required = Checkbox(id='ansible_variable_required')
+    validator_type = SatSelect(id='ansible_variable_validator_type')
+    validator_rule = TextInput(id='ansible_variable_validator_rule')
+    # the below attribute_order widget throws an error when a string value is passed to it
+    attribute_order = TextInput(id='order')
+    merge_overrides = Checkbox(id='ansible_variable_merge_overrides')
+    merge_default = Checkbox(id='ansible_variable_merge_default')
+    avoid_duplicates = Checkbox(id='ansible_variable_avoid_duplicates')
+    add_matcher_button = Text("//a[contains(@class, 'add_nested_fields')]")
+    attribute_type = SatSelect("//div[@class='matcher-group']/select[1]")
+    # the below attribute_value widget is not locating the element correctly as currently written
+    attribute_value = Text("//div[@class='matcher-group']/input[1]")
```

Comment:
These will be a little bit difficult to handle because there could be potential multiple `Attribute type` when you click `Add Matcher`.  One shouldn't give much issue when it's trying to search for the one available field though.  Try `@class='matcher_value'` and see if that works.  if it doesn't, you may need to set the `locator` to that or pass in the `input` one.

Reply:
I've been working on this today. Still unable to get the `attribute_value` field working. I've tried the following XPath expressions:
- `TextInput("//div[@class='matcher-group']/input[1]")`
- `TextInput("//input[contains(@class, 'matcher_value')")`
- `TextInput("//input[@class='matcher_value']")`
- several other minor variations on the above.

All of these throw a `NoSuchElementException` when my test reaches the `attribute_value` variable. Weirdly, the current XPath expression in the PR (`Text("//div[@class='matcher-group']/input[1]"`) does throw the same exception. However, `Text` objects do not have a `fill` method, and changing `Text` to `TextInput` brings back the exception. I do not have an explanation for this discrepancy in finding the element.

At this point, I am inclined to relegate tests involving matchers to manual tests for the time being, as this issue is preventing us from automating Ansible roles and variables tests that do not involve matchers. @latran @mshriver I would be interested to hear your perspectives on this idea.

Reply:
@mshriver @latran I was able to spend some more time on this today. One of the things that I tried was a much more verbose XPath expression to point to this input field:
`TextInput("//div[@class='matcher-parent']/fieldset[2]/div[1]/table[1]/tbody[1]/tr[1]/td[1]/div[1]/input[1]")`
I verified in a browser console that this expression does indeed point to the intended text input field. However, my test still throws a `NoSuchElementException` when it attempts to use this input field. For reference, the test I'm using is here [1]. 

I could really use some help figuring out why I'm getting this error with an apparently correct XPath expression. 

[1] https://github.com/synkd/robottelo/blob/cf480c9c8c90905acf4912a04f17d27680e88cee/tests/foreman/ui/test_ansible.py#L78

Reply:
Incidentally, using XPath expressions that point to specific table rows is a possible solution to the issue of multiple `attribute_type` variables that Lai inititally raised in this thread. I'm thinking of something like:
```
attribute_types = [
    TextInput("//div[@class='matcher-parent']/fieldset[2]/div[1]/table[1]/tbody[1]/tr[1]/td[1]/div[1]/input[1]"),
    TextInput("//div[@class='matcher-parent']/fieldset[2]/div[1]/table[1]/tbody[1]/tr[2]/td[1]/div[1]/input[1]"),
]
```
The list could contain a number items equal to the largest number of Ansible Variable matchers needed by a test.

Reply:
@synkd I think you need to refactor this code. I think adding a table view makes more sense and attaching that to Main View e.g.

```
class MatcherView(View):
    ROOT = '//h2[text()="Specify Matchers  "]/..'
    add_matcher = Text("//a[contains(@class, 'add_nested_fields')]")
    table = SatTable(
        locator=".//table[@class='table white-header']",
        column_widgets={
            'Attribute type': Select(".//select"),
            'Value': TextInput(".//textarea"),
            'Actions': Text(".//a"),
        },
    ) 
```

Reply:
@omkarkhatavkar Thank you for looking at this. Using a table widget does make sense, since that's how that part of the HTML is structured. However, I'm not clear on what the advantage would be to adding a table view, as opposed to adding a table to the existing `NewAnsibleVariableView`. Could you please explain your thinking there? Is it to separate filling the values in the `Specify Matchers` section from filling the data on the rest of the page?

Reply:
Added more details implementation here https://github.com/synkd/airgun/pull/1, you can modify this behaviour if you like

Reply:
Thanks to some amazing collab from @omkarkhatavkar, this is now working as expected. I need to do some additional testing to verify that Omkar's changes to the `CustomParameter` widget do not affect other tests, and  need to resolve an issue with my delete method in the Ansible Variable entity where the confirmation dialog is not being handled properly. Once these tasks are completed, this PR will be ready for re-review.

Reply:
One additional note: I had to recreate separate methods for creating a variable with and without overrides. When the `Override` checkbox is selected but no values are entered into the section of the form enabled by that checkbox, Satellite throws an error when the form is submitted. This does not strike me as a product bug, so I've added separate methods for those two scenarios (with and without overrides) to the entity.